stablecoin-v3/src/lib.rs

```diff
@@ -0,0 +1,286 @@
+        ERROR_PRICE_AGGREGATOR_WRONG_ADDRESS, ERROR_SAME_TOKENS, ERROR_SLIPPAGE_EXCEEDED,
+        ERROR_SWAP_NOT_ENABLED,
+    },
+    events::SwapEvent,
+};
+
+const MEDIAN_POOL_DELTA: u64 = 100_000_000;
+const PERCENTAGE: u64 = 100_000;
+
+#[elrond_wasm::contract]
+pub trait StablecoinV3:
+    virtual_liquidity_pools::VLPModule + config::ConfigModule + events::EventsModule
+{
+    #[init]
+    fn init(&self, price_aggregator_address: ManagedAddress, pool_recovery_period: u64) {
+        require!(
+            !price_aggregator_address.is_zero(),
+            ERROR_PRICE_AGGREGATOR_WRONG_ADDRESS
+        );
+        self.price_aggregator_address()
+            .set(&price_aggregator_address);
+        self.pool_recovery_period().set(pool_recovery_period);
+        self.state().set(&State::Inactive);
+    }
+
+    #[only_owner]
+    #[payable("EGLD")]
+    #[endpoint(registerStablecoin)]
+    fn register_stablecoin(
+        &self,
+        token_display_name: ManagedBuffer,
+        token_ticker: ManagedBuffer,
+        num_decimals: usize,
+    ) {
+        require!(!self.is_state_active(), ERROR_ACTIVE);
+        let payment_amount = self.call_value().egld_value();
+        self.stablecoin().issue_and_set_all_roles(
+            payment_amount,
+            token_display_name,
+            token_ticker,
+            num_decimals,
+            None,
+        );
+    }
+
+    #[only_owner]
+    #[payable("*")]
+    #[endpoint(deployStablecoin)]
+    fn deploy_stablecoin(
+        &self,
+        collateral_token_id: TokenIdentifier,
+        stablecoin_token_id: TokenIdentifier,
+        collateral_token_ticker: ManagedBuffer,
+        stablecoin_token_ticker: ManagedBuffer,
+        spread_fee_min_percent: BigUint,
+    ) -> BigUint {
+        let (payment_amount, token_in) = self.call_value().payment_token_pair();
+
+        require!(collateral_token_id.is_esdt(), ERROR_NOT_AN_ESDT);
+        require!(stablecoin_token_id.is_esdt(), ERROR_NOT_AN_ESDT);
+        require!(
+            collateral_token_id != stablecoin_token_id,
+            ERROR_SAME_TOKENS
+        );
+        require!(token_in == collateral_token_id, ERROR_BAD_PAYMENT_TOKENS);
+        require!(self.base_pool().is_empty(), ERROR_ALREADY_DEPLOYED);
+
+        self.spread_fee_min_percent().set(spread_fee_min_percent);
+        self.stablecoin().set_token_id(&stablecoin_token_id);
+        self.collateral_token_id().set(&collateral_token_id);
+        self.token_ticker(&collateral_token_id)
+            .set(collateral_token_ticker);
+        self.token_ticker(&stablecoin_token_id)
+            .set(stablecoin_token_ticker);
+        self.pool_delta().set(&BigUint::from(MEDIAN_POOL_DELTA));
+
+        let caller = self.blockchain().get_caller();
+        let collateral_price = self.get_exchange_rate(&collateral_token_id, &stablecoin_token_id);
+        let payment_value_denominated = &payment_amount * &collateral_price;
```

Comment:
The oracle will give you `collateral_price` as denominated already, so to get the actual `payment_value_denominated`, you should divide by `10^decimals` after the multiplication.

For example, if the collateral is EGLD, and its price is $100, then the oracle will return 100*10^18 (assuming all prices have 18 decimals). Assuming the user brought 10 EGLD, your calculation would result in:
payment_value_denominated = (10 * 10^18) * (100 * 10^18), so you'd need to divide by 10^18.